### DISCUSSION TITLE
Use github.com/benbjohnson/clock to supply time functionality to Grafana

### DISCUSSION THREAD

**Discussion by joeblubaugh** [grafana/grafana#46822] [General]
Grafana currently uses the `time` package directly when access to computer time is required. This is difficult to use in tests, especially tests that involve code with tickers or other interval-dependent functionality.

github.com/benbjohnson/clock provides a `clock.Clock` type that implements the `time` package functions like `Now()` and provides a ticker. In the "standard" configuration is calls `time` directly, but you can also use a mocked Clock in tests, where you can manipulate the current time to specific dates, advance the clock to trigger timers, etc. This can make time-dependent tests much more deterministic.

Using a package like this can be added incrementally, instantiating a clock object where needed, rather than whole-sale through the project.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
